driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -328,4 +332,16 @@ protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out)
             }
         }
     }
+
+    @ChannelHandler.Sharable
+    static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+            ProtocolVersion protocolVersion = frame.header.version;
+            // Check for the protocol version, if the downgrade has occurred during renegotiation
+            if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
+                frame.header.flags.add(Header.Flag.USE_BETA);
```

Comment:
> Wouldn't it be slightly simpler if we didn't allow auto-negotiation with beta versions?
If you want to use a beta, you would have to specify the version as well (we could even have an alternate method `Cluster.Builder#withBetaProtocolVersion(ProtocolVersion)`).

Yes we could, and that was close to what @ifesdjeen had in his early versions. I found it a bit unfriendly as that would throw an `UnsupportedProtocolVersionException` back to the user in case A1 in my table above. So I suggested instead the opposite as I thought it would be interesting for users to simply declare that they are willing to use beta versions (and cope with the consequences), without actually specifying the version. This way renegotiation would take care of downgrading the version gracefully if need be.

But long story short, I'm fine if we decide to go down the path you suggested, specially if it simplifies the code.